components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.cc

```diff
@@ -0,0 +1,112 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_ads/content/browser/search_result_ad/search_result_ad_redirect_throttle.h"
+
+#include <utility>
+
+#include "brave/components/brave_ads/common/search_result_ad_util.h"
+#include "brave/components/brave_ads/content/browser/search_result_ad/search_result_ad_service.h"
+#include "brave/components/brave_search/common/brave_search_utils.h"
+#include "components/sessions/content/session_tab_helper.h"
+#include "services/network/public/cpp/request_mode.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "third_party/blink/public/mojom/loader/resource_load_info.mojom-shared.h"
+#include "url/gurl.h"
+
+namespace brave_ads {
+
+namespace {
+
+SessionID GetTabId(content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  SessionID tab_id = sessions::SessionTabHelper::IdForTab(web_contents);
+  content::RenderFrameHost* opener_rfh = web_contents->GetOpener();
+  if (!opener_rfh) {
+    opener_rfh = web_contents->GetOriginalOpener();
+  }
+  if (opener_rfh) {
+    content::WebContents* original_web_contents =
+        content::WebContents::FromRenderFrameHost(opener_rfh);
+    tab_id = sessions::SessionTabHelper::IdForTab(original_web_contents);
+  }
+
+  return tab_id;
+}
+
+}  // namespace
+
+// static
+std::unique_ptr<SearchResultAdRedirectThrottle>
+SearchResultAdRedirectThrottle::MaybeCreateThrottleFor(
+    SearchResultAdService* search_result_ad_service,
+    const network::ResourceRequest& request,
+    content::WebContents* web_contents) {
+  DCHECK(web_contents);
+
+  if (!search_result_ad_service || !request.request_initiator ||
+      !request.has_user_gesture || !request.is_outermost_main_frame ||
+      request.resource_type !=
+          static_cast<int>(blink::mojom::ResourceType::kMainFrame)) {
+    return nullptr;
+  }
+
+  if (!brave_search::IsAllowedHost(request.request_initiator->GetURL()) ||
+      !IsSearchResultAdClickedConfirmationUrl(request.url)) {
+    return nullptr;
+  }
+
+  SessionID tab_id = GetTabId(web_contents);
```

Comment:
what are you using tab_id for?

Reply:
I'm not following why that matters, but in any case there is a memory leak because there is no guarantee that `DocumentOnLoadCompletedInPrimaryMainFrame` will be called (tab close or redirect).

Reply:
you could clear on `~AdsTabHelper` which would at least take care of the memory leak

Reply:
also since you are clearing in DidFinishNavigation you are already limiting it to the lifetime of the navigation because any pending callbacks get cleared in ResetState

Reply:
yeah, the clear code is called in `AdsTabHelper::WebContentsDestroyed`